### ISSUE TITLE
[epoll] Reactor does not register interest in some cases. Issue introduced after read_with/write_with commit.

### ISSUE THREAD

**Issue by oblique** [stjepang/smol#161]
I updated to the latest smol on a pet project that uses UnixStream for communication and I notice that in some cases communication was getting stuck on `epoll_wait`.

I did bisect on smol and the first bad commit is 76aa1ea654a322d0abaf4291dde2eb5d89c50205.

I also created a minimal reproducible example: https://github.com/oblique/smol-epoll-issue

Something that I noticed is that if I switch from UnixStream to TcpStream, the bug is not reproducible anymore.

To reproduce it do:

* Clone `smol-epoll-issue`
* Run: `cargo run --bin server`
* Run: `cargo run --bin client`
* Client will start printing some text
* After some seconds text printing will get stuck

**Comment by sopium** [comment]
Thanks! This is an important case (where both read and write would block) that should probably be added to the test suite of smol itself.

@stjepang I believe we need to `reregister` the interest when previously we are interested in both readable and writable, but only one of them is emitted, because we are still interested in the other:

(Maybe this will fix #136 too.)

```diff
diff --git a/src/reactor.rs b/src/reactor.rs
index c5cd338..6ba4b65 100644
--- a/src/reactor.rs
+++ b/src/reactor.rs
@@ -293,6 +293,15 @@ impl ReactorLock<'_> {
                             if ev.writable {
                                 ready.append(&mut wakers.writers);
                             }
+
+                            if !(wakers.writers.is_empty() && wakers.readers.is_empty()) {
+                                self.reactor.sys.reregister(
+                                    source.raw,
+                                    source.key,
+                                    !wakers.readers.is_empty(),
+                                    !wakers.writers.is_empty(),
+                                )?;
+                            }
                         }
                     }

```

**Comment by oblique** [comment]
I tried this patch and it fixes the problem.
